testsuite/tests/syntactic-arity/comparative_alloc.ml

```diff
@@ -0,0 +1,121 @@
+(* TEST
+ flags = "-w +A-70";
+ setup-ocamlopt.byte-build-env;
+ ocamlopt.byte;
+ run;
+ check-program-output;
+*)
+
+(* Check that the runtime arity of a function (i.e., its 'fast path' for
+   runtime application) matches its syntactic arity (i.e., the number
+   of arguments appearing directly following [fun]).
+*)
+
+type (_, _) raw_arity =
+  | One : (int -> 'ret, 'ret) raw_arity
+  | Succ : ('f, 'ret) raw_arity -> (int -> 'f, 'ret) raw_arity
+
+let rec numeric_arity : type f ret. (f, ret) raw_arity -> int =
+  fun arity ->
+    match arity with
+    | One -> 1
+    | Succ arity -> numeric_arity arity + 1
+
+let rec apply : type f ret. (f, ret) raw_arity -> f -> int -> ret =
+  fun arity f arg ->
+    match arity with
+    | One -> f arg
+    | Succ arity -> apply arity (f arg) arg
+
+type 'a arity =
+  | Tupled
+  | Curried : ('a, _) raw_arity -> 'a arity
+
+type packed_raw_arity = Packed_raw_arity : _ raw_arity -> packed_raw_arity
+type packed_arity = Packed_arity : _ arity -> packed_arity
+
+let arity_description (type a) (arity : a arity) =
+  match arity with
+  | Tupled -> "tupled fun"
+  | Curried arity -> Printf.sprintf "%d-ary fun" (numeric_arity arity)
+
+(* [runtime_arity] depends on representation details of functions and
+   is subject to change.
+*)
+let runtime_arity (f : 'a -> 'b) : ('a -> 'b) arity =
+  let clos_info = Obj.raw_field (Obj.repr f) 1 in
```

Comment:
That `1`, and the shifting logic below, makes me rather nervous. It's in general possible that the structure of closures will change. I guess it's probably fine, though, because the person making that change will run this test, see what's gone wrong, and be well-placed to fix it. Still, I think having a `let extract_arity_from_closure (f_obj : Obj.t) = ...` at the top of the file will only help.

Reply:
Good idea, I've done that.